core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,184 @@
+    assertTrue(memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID)
+    assertTrue(memberEpoch != JoinGroupRequest.UNKNOWN_GENERATION_ID)
+
+    createTopic(topic, 1)
+
+    var producerIdAndEpoch: ProducerIdAndEpoch = null
+    // Wait until ALLOCATE_PRODUCER_ID request finished
+    TestUtils.waitUntilTrue(() =>
+      try {
+        producerIdAndEpoch = initProducerId(
+          transactionalId = transactionalId,
+          producerIdAndEpoch = ProducerIdAndEpoch.NONE,
+          expectedError = Errors.NONE)
+        true
+      } catch {
+        case _: Throwable => false
+      }, "initProducerId request failed"
+    )
+
+    addOffsetsToTxn(
+      groupId = groupId,
+      producerId = producerIdAndEpoch.producerId,
+      producerEpoch = producerIdAndEpoch.epoch,
+      transactionalId = transactionalId
+    )
+
+    def verifyTxnCommitAndFetch(
+      groupId: String,
+      memberId: String,
+      generationId: Int,
+      offset: Long,
+      version: Short
+    ): Unit = {
+      commitTxnOffset(
+        groupId = groupId,
+        memberId = memberId,
+        generationId = generationId,
+        producerId = producerIdAndEpoch.producerId,
+        producerEpoch = producerIdAndEpoch.epoch,
+        transactionalId = transactionalId,
+        topic = topic,
+        partition = partition,
+        offset = offset,
+        expectedError = Errors.NONE,
+        version = version
+      )
+
+      // Send writeTxnMarker to force transaction from pending to completed
+      // hence we can retrieve offsets from OFFSET_FETCH request
+      val writableTxnMarkersResult = writeTxnMarkers(
+        Collections.singletonList(new WritableTxnMarker()
+          .setTopics(Collections.singletonList(new WritableTxnMarkerTopic()
+            .setName(Topic.GROUP_METADATA_TOPIC_NAME)
+            .setPartitionIndexes(Collections.singletonList(0))))
+          .setProducerId(producerIdAndEpoch.producerId)
+          .setProducerEpoch(producerIdAndEpoch.epoch)
+          .setCoordinatorEpoch(0.toShort)
+          .setTransactionResult(true))
+      )
+      assertEquals(1, writableTxnMarkersResult.size())
+      val writableTxnMarkerTopic = writableTxnMarkersResult.asScala.head.topics().asScala.head
+      assertEquals(Topic.GROUP_METADATA_TOPIC_NAME, writableTxnMarkerTopic.name())
+      assertEquals(1, writableTxnMarkerTopic.partitions().size())
+      assertEquals(Errors.NONE.code(), writableTxnMarkerTopic.partitions().asScala.head.errorCode())
+
+      val fetchOffsetsResp = fetchOffsets(
+        groups = Map(groupId -> List(new TopicPartition(topic, partition))),
+        requireStable = true,
+        version = ApiKeys.OFFSET_FETCH.latestVersion()
+      )
+      val groupIdRecord = fetchOffsetsResp.find(_.groupId() == groupId).head
+      val topicRecord = groupIdRecord.topics().asScala.find(_.name() == topic).head
+      val partitionRecord = topicRecord.partitions().asScala.find(_.partitionIndex() == partition).head
```

Comment:
nit: In Scala, we omit the `()` for getters: `topicRecord.partitions.asScala.find(_.partitionIndex == partition).head`. They are other cases in this test. Could you please check them?

Reply:
Thanks for reviewing again, @dajac. All comments are addressed in the latest commit, please take a look when you are available.